File: core/src/nl/t64/game/rpg/components/quest/QuestGraph.java
package nl.t64.game.rpg.components.quest;

import lombok.Getter;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.loot.Loot;
import nl.t64.game.rpg.constants.Constant;
import nl.t64.game.rpg.screens.world.conversation.ConversationSubject;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static java.util.function.Predicate.not;


@Getter
public class QuestGraph {

    private static final QuestState DEFAULT_STATE = QuestState.UNKNOWN;

    String id;
    QuestState currentState;
    boolean isFailed;
    private String title;
    private String entityId;
    private String summary;
    private String linkedWith;
    private Map<String, QuestTask> tasks;

    private QuestGraph() {
        this.currentState = DEFAULT_STATE;
        this.isFailed = false;
    }

    public String toString() {
        if (isFailed) {
            return "x  " + title;
        } else if (currentState.equals(QuestState.FINISHED)) {
            return "v  " + title;
        } else if (currentState.equals(QuestState.UNCLAIMED)) {
            return "o   " + title;
        } else {
            return "     " + title;
        }
    }

    public List<QuestTask> getAllTasks() {
        return tasks.entrySet().stream()
                    .sorted(Map.Entry.comparingByKey())
                    .map(Map.Entry::getValue)
                    .collect(Collectors.toUnmodifiableList());
    }

    public void setTaskComplete(String taskId) {
        tasks.get(taskId).setComplete();
    }

    public boolean isTaskComplete(String taskId) {
        return tasks.get(taskId).isComplete();
    }

    public boolean isFinished() {
        return currentState.equals(QuestState.FINISHED);
    }

    public void handleAccept(Consumer<String> continueConversation, ConversationSubject observers) {
        handleTolerate(observers);
        if (doesReturnMeetDemand()) {
            continueConversation.accept(Constant.PHRASE_ID_QUEST_IMMEDIATE_SUCCESS);
        } else {
            continueConversation.accept(Constant.PHRASE_ID_QUEST_ACCEPT);
        }
    }

    public void handleTolerate(ConversationSubject observers) {
        know();
        accept(observers);
    }

    public void handleReceive(ConversationSubject observers) {
        know();
        Loot receive = getAllQuestTasks().stream()
                                         .filter(questTask -> questTask.type.equals(QuestType.ITEM_DELIVERY))
                                         .findFirst()
                                         .map(questTask -> new Loot(questTask.getTarget()))
                                         .orElseThrow();
        observers.notifyShowReceiveDialog(receive);
    }

    public void handleCheckIfLinkedIsKnown(String phraseId, Consumer<String> continueConversation) {
        if (linkedWith != null
            && !Utils.getGameData().getQuests().getQuestById(linkedWith).getCurrentState().equals(QuestState.UNKNOWN)) {
            continueConversation.accept(Constant.PHRASE_ID_QUEST_LINKED);
        } else {
            continueConversation.accept(phraseId);
        }
    }

    public void handleCheckIfAccepted(String phraseId,
                                      Consumer<String> continueConversation,
                                      Consumer<String> endConversation) {
        if (currentState.equals(QuestState.ACCEPTED)) {
            continueConversation.accept(Constant.PHRASE_ID_QUEST_DELIVERY);
        } else {
            endConversation.accept(phraseId);
        }
    }

    public void handleCheckIfAcceptedInventory(String taskId, String phraseId,
                                               Consumer<String> continueConversation,
                                               Consumer<String> endConversation) {
        if (currentState.equals(QuestState.ACCEPTED)
            && tasks.get(taskId).hasTargetInInventory()) {
            continueConversation.accept(Constant.PHRASE_ID_QUEST_DELIVERY);
        } else {
            endConversation.accept(phraseId);
        }
    }

    public void handleReturn(Consumer<String> continueConversation) {
        if (doesReturnMeetDemand()) {
            continueConversation.accept(Constant.PHRASE_ID_QUEST_SUCCESS);
        } else {
            continueConversation.accept(Constant.PHRASE_ID_QUEST_NO_SUCCESS);
        }
    }

    public void handleReward(Consumer<String> endConversation, ConversationSubject observers) {
        if (currentState.equals(QuestState.ACCEPTED)) {
            takeDemands();
            unclaim();
            getAllQuestTasks().forEach(QuestTask::forceFinished);
            observers.notifyShowMessageTooltip(
                    "Quest completed:" + System.lineSeparator() + System.lineSeparator() + title);
        }
        Loot reward = Utils.getGameData().getLoot().getLoot(id);
        reward.removeBonus();
        Optional<String> levelUpMessage = partyGainXp(reward, observers);
        if (reward.isTaken()) {
            finish();
            endConversation.accept(Constant.PHRASE_ID_QUEST_FINISHED);
            levelUpMessage.ifPresent(observers::notifyShowLevelUpDialog);
        } else {
            observers.notifyShowRewardDialog(reward, levelUpMessage.orElse(null));
        }
    }

    public void handleFail(ConversationSubject observers) {
        isFailed = true;
        observers.notifyShowMessageTooltip("Quest failed:" + System.lineSeparator() + System.lineSeparator() + title);
    }

    public void know() {
        switch (currentState) {
            case KNOWN:
                break;
            case UNKNOWN:
                currentState = QuestState.KNOWN;
                break;
            default:
                throw new IllegalStateException("Only quest UNKNOWN can be KNOWN.");
        }
    }

    public void accept(ConversationSubject observers) {
        if (currentState.equals(QuestState.KNOWN)) {
            currentState = QuestState.ACCEPTED;
            observers.notifyShowMessageTooltip("New quest:" + System.lineSeparator() + System.lineSeparator() + title);
        } else {
            throw new IllegalStateException("Only quest KNOWN can be ACCEPTED.");
        }
    }

    public void unclaim() {
        if (currentState.equals(QuestState.ACCEPTED)) {
            currentState = QuestState.UNCLAIMED;
        } else {
            throw new IllegalStateException("Only quest ACCEPTED can be UNCLAIMED.");
        }
    }

    public void finish() {
        if (currentState.equals(QuestState.UNCLAIMED)) {
            currentState = QuestState.FINISHED;
        } else {
            throw new IllegalStateException("Only quest UNCLAIMED can be FINISHED.");
        }
    }

    private void takeDemands() {
        getAllQuestTasks().stream()
                          .filter(questTask -> questTask.type.equals(QuestType.FETCH))
                          .forEach(QuestTask::removeTargetFromInventory);
    }

    private Optional<String> partyGainXp(Loot reward, ConversationSubject observers) {
        if (reward.isXpGained()) {
            return Optional.empty();
        }
        StringBuilder levelUpMessage = new StringBuilder();
        Utils.getGameData().getParty().getAllHeroes().forEach(hero -> hero.gainXp(reward.getXp(), levelUpMessage));
        observers.notifyShowMessageTooltip(String.format("+ %s XP", reward.getXp()));
        reward.clearXp();
        String finalMessage = levelUpMessage.toString().strip();
        if (finalMessage.isEmpty()) {
            return Optional.empty();
        } else {
            return Optional.of(finalMessage);
        }
    }

    private boolean doesReturnMeetDemand() {
        return getAllQuestTasks().stream()
                                 .filter(not(QuestTask::isOptional))
                                 .allMatch(QuestTask::isCompleteForReturn);
    }

    private List<QuestTask> getAllQuestTasks() {
        return List.copyOf(tasks.values());
    }

}

File: core/src/nl/t64/game/rpg/components/quest/QuestGraph.kt
package nl.t64.game.rpg.components.quest

import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.components.loot.Loot
import nl.t64.game.rpg.constants.Constant
import nl.t64.game.rpg.screens.world.conversation.ConversationSubject


private val DEFAULT_STATE = QuestState.UNKNOWN

class QuestGraph(
    val title: String = "",
    val entityId: String = "",
    val summary: String = "",
    val isHidden: Boolean = false,
    val linkedWith: String? = null,
    val tasks: Map<String, QuestTask> = emptyMap()
) {
    lateinit var id: String
    var currentState: QuestState = DEFAULT_STATE
    var isFailed: Boolean = false

    override fun toString(): String {
        return when {
            isFailed -> "x  $title"
            currentState == QuestState.FINISHED -> "v  $title"
            currentState == QuestState.UNCLAIMED -> "o   $title"
            else -> "     $title"
        }
    }

    fun isCurrentStateEqualOrHigherThan(questState: QuestState): Boolean = currentState.isEqualOrHigherThan(questState)
    fun isCurrentStateEqualOrLowerThan(questState: QuestState): Boolean = currentState.isEqualOrLowerThan(questState)

    fun getAllTasks(): Array<QuestTask> = tasks.entries
        .sortedBy { it.key }
        .map { it.value }
        .toTypedArray()

    fun setTaskComplete(taskId: String) {
        tasks[taskId]!!.setComplete()
    }

    fun isTaskComplete(taskId: String?): Boolean {
        if (taskId == null) return false
        return tasks[taskId]!!.isComplete
    }

    fun handleAccept(continueConversation: (String) -> Unit, observers: ConversationSubject) {
        handleTolerate(observers)
        val phraseId =
            if (doesReturnMeetDemand()) Constant.PHRASE_ID_QUEST_IMMEDIATE_SUCCESS else Constant.PHRASE_ID_QUEST_ACCEPT
        continueConversation.invoke(phraseId)
    }

    fun handleTolerate(observers: ConversationSubject) {
        know()
        accept(observers)
    }

    fun handleReceive(observers: ConversationSubject) {
        know()
        val receiveLoot = getAllQuestTasks()
            .filter { it.type == QuestType.ITEM_DELIVERY }
            .map { Loot(it.target as MutableMap<String, Int>) }
            .first()
        observers.notifyShowReceiveDialog(receiveLoot)
    }

    fun handleCheckIfLinkedIsKnown(phraseId: String, continueConversation: (String) -> Unit) {
        val newPhraseId =
            if (linkedWith != null && gameData.quests.isCurrentStateEqualOrHigherThan(linkedWith, QuestState.KNOWN)) {
                Constant.PHRASE_ID_QUEST_LINKED
            } else phraseId
        continueConversation.invoke(newPhraseId)
    }

    fun handleCheckIfAccepted(
        phraseId: String,
        continueConversation: (String) -> Unit,
        endConversation: (String) -> Unit
    ) {
        if (currentState == QuestState.ACCEPTED) {
            continueConversation.invoke(Constant.PHRASE_ID_QUEST_DELIVERY)
        } else {
            endConversation.invoke(phraseId)
        }
    }

    fun handleCheckIfAcceptedInventory(
        taskId: String,
        phraseId: String,
        continueConversation: (String) -> Unit,
        endConversation: (String) -> Unit
    ) {
        if (currentState == QuestState.ACCEPTED && tasks[taskId]!!.hasTargetInInventory()) {
            continueConversation.invoke(Constant.PHRASE_ID_QUEST_DELIVERY)
        } else {
            endConversation.invoke(phraseId)
        }
    }

    fun handleReturn(continueConversation: (String) -> Unit) {
        val phraseId =
            if (doesReturnMeetDemand()) Constant.PHRASE_ID_QUEST_SUCCESS else Constant.PHRASE_ID_QUEST_NO_SUCCESS
        continueConversation.invoke(phraseId)
    }

    fun handleAcceptOrReturn(continueConversation: (String) -> Unit, observers: ConversationSubject) {
        if (isCurrentStateEqualOrLowerThan(QuestState.KNOWN)) {
            handleAccept(continueConversation, observers)
        } else if (currentState == QuestState.ACCEPTED) {
            handleReturn(continueConversation)
        } else {
            throw IllegalStateException("You've messed up the conversation quest flow.")
        }
    }

    fun handleReward(endConversation: (String) -> Unit, observers: ConversationSubject) {
        if (currentState == QuestState.ACCEPTED) {
            handleRewardPart1(observers)
        }
        if (currentState == QuestState.UNCLAIMED) {
            handleRewardPart2(observers, endConversation)
        } else {
            throw IllegalStateException("You've messed up the conversation quest flow.")
        }
    }

    fun handleFail(observers: ConversationSubject) {
        isFailed = true
        observers.notifyShowMessageTooltip("Quest failed:" + System.lineSeparator() + System.lineSeparator() + title)
    }

    fun know() {
        when (currentState) {
            QuestState.KNOWN -> {
            }
            QuestState.UNKNOWN -> currentState = QuestState.KNOWN
            else -> throw IllegalStateException("Only quest UNKNOWN can be KNOWN.")
        }
    }

    fun accept(observers: ConversationSubject) {
        if (currentState == QuestState.KNOWN) {
            currentState = QuestState.ACCEPTED
            if (!isHidden) observers.notifyShowMessageTooltip("New quest:" + System.lineSeparator() + System.lineSeparator() + title)
        } else {
            throw IllegalStateException("Only quest KNOWN can be ACCEPTED.")
        }
    }

    fun unclaim() {
        if (currentState == QuestState.ACCEPTED) {
            currentState = QuestState.UNCLAIMED
        } else {
            throw IllegalStateException("Only quest ACCEPTED can be UNCLAIMED.")
        }
    }

    fun finish() {
        if (currentState == QuestState.UNCLAIMED) {
            currentState = QuestState.FINISHED
        } else {
            throw IllegalStateException("Only quest UNCLAIMED can be FINISHED.")
        }
    }

    private fun handleRewardPart1(observers: ConversationSubject) {
        takeDemands()
        unclaim()
        getAllQuestTasks().forEach { it.forceFinished() }
        if (!isHidden) observers.notifyShowMessageTooltip("Quest completed:" + System.lineSeparator() + System.lineSeparator() + title)
    }

    private fun handleRewardPart2(observers: ConversationSubject, endConversation: (String) -> Unit) {
        val reward = gameData.loot.getLoot(id)
        reward.removeBonus()
        val levelUpMessage = partyGainXp(reward, observers)
        if (reward.isTaken()) {
            finish()
            endConversation.invoke(Constant.PHRASE_ID_QUEST_FINISHED)
            levelUpMessage?.let { observers.notifyShowLevelUpDialog(it) }
        } else {
            observers.notifyShowRewardDialog(reward, levelUpMessage)
        }
    }

    private fun takeDemands() {
        getAllQuestTasks()
            .filter { it.type == QuestType.FETCH_ITEM }
            .forEach { it.removeTargetFromInventory() }
    }

    private fun partyGainXp(reward: Loot, observers: ConversationSubject): String? {
        if (reward.isXpGained()) {
            return null
        }
        val levelUpMessage = StringBuilder()
        gameData.party.gainXp(reward.xp, levelUpMessage)
        observers.notifyShowMessageTooltip("+ ${reward.xp} XP")
        reward.clearXp()
        return levelUpMessage.toString().trim().ifEmpty { null }
    }

    private fun doesReturnMeetDemand(): Boolean {
        return getAllQuestTasks()
            .filter { !it.isOptional }
            .all { it.isCompleteForReturn() }
    }

    private fun getAllQuestTasks(): List<QuestTask> = ArrayList(tasks.values)

}
